From 36245ad8a1ffd19899f0dc7854c5787389f0f3ec Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Wed, 8 Jun 2022 13:13:38 +0000
Subject: [PATCH] xen/arm: Implement MPAM sysctls for L3 cache

This commit implements the MPAM sysctls (XEN_SYSCTL_psr_cmt_op and
XEN_SYSCTL_psr_alloc) provided by xl-psr for L3 cache, which allows
user to get system level MPAM information of the hardware platform,
namely max-supported PARTIDs, max-supported PMGs, max-supported
cache portion bitmap width, and L3 cache size of the platform,
through the "xl psr-hwinfo" command.

Related helpers are implemented in mpam.c for conveniency of the
callers.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-3763
---
 xen/arch/arm/include/asm/mpam.h |  6 ++++
 xen/arch/arm/mpam.c             | 45 ++++++++++++++++++++++++++--
 xen/arch/arm/sysctl.c           | 52 +++++++++++++++++++++++++++++++++
 3 files changed, 101 insertions(+), 2 deletions(-)

diff --git a/xen/arch/arm/include/asm/mpam.h b/xen/arch/arm/include/asm/mpam.h
index 286fd5dd8851..6528d89b462d 100644
--- a/xen/arch/arm/include/asm/mpam.h
+++ b/xen/arch/arm/include/asm/mpam.h
@@ -418,6 +418,8 @@ struct mpam_domain {
 extern struct xen_arch_domctl_mpam dom0_mpam_config;
 extern bool mpam_enabled;
 
+extern void mpam_get_cache_size(enum mpam_class_types cache_type,
+                                unsigned int cache_level, unsigned int *size);
 extern void mpam_postsmp_setup(void);
 extern void mpam_presmp_setup(void);
 extern void mpam_restore_state(struct vcpu *n);
@@ -429,6 +431,10 @@ extern void mpam_set_vcpu_partid_pmg(struct vcpu *vcpu,
                                      uint8_t pmg_d, uint8_t pmg_i);
 extern void mpam_update_domain(struct domain *d);
 extern int mpam_discovery_cpu_online(unsigned int cpu);
+extern int mpam_get_hwinfo(enum mpam_class_types type, unsigned int level,
+                           unsigned int *val);
+extern uint16_t mpam_get_system_max_partid(void);
+extern uint8_t mpam_get_system_max_pmg(void);
 
 #endif /* __ARM_MPAM_H__ */
 
diff --git a/xen/arch/arm/mpam.c b/xen/arch/arm/mpam.c
index 7f18add93548..8f2074fcec88 100644
--- a/xen/arch/arm/mpam.c
+++ b/xen/arch/arm/mpam.c
@@ -1419,8 +1419,8 @@ static int __init mpam_probe_cache_info(void)
  * Get MPAM hwinfo based on the input resource type.
  * Currently assume MPAM hwinfo is the same across different clusters.
  */
-static int mpam_get_hwinfo(enum mpam_class_types type, unsigned int level,
-                           unsigned int *val)
+int mpam_get_hwinfo(enum mpam_class_types type, unsigned int level,
+                    unsigned int *val)
 {
     struct mpam_class *class;
 
@@ -1579,6 +1579,47 @@ void mpam_set_vcpu_partid_pmg(struct vcpu *vcpu,
     vcpu->arch.mpam1_el1 = regval;
 }
 
+uint16_t mpam_get_system_max_partid(void)
+{
+    return ACCESS_ONCE(mpam_partid_max);
+}
+
+uint8_t mpam_get_system_max_pmg(void)
+{
+    return ACCESS_ONCE(mpam_pmg_max);
+}
+
+void mpam_get_cache_size(enum mpam_class_types cache_type,
+                         unsigned int cache_level, unsigned int *size)
+{
+    struct mpam_class *cache;
+
+    spin_lock(&mpam_list_lock);
+    /* We do not really care of the cache level of SLC */
+    if ( cache_type == MPAM_CLASS_SLC )
+    {
+        list_for_each_entry(cache, &mpam_classes, classes_list)
+        {
+            if ( cache->type == MPAM_CLASS_SLC )
+                break;
+        }
+    }
+    else
+    {
+        cache = mpam_class_get(cache_level, cache_type, false);
+        if ( IS_ERR(cache) )
+        {
+            printk(XENLOG_ERR "MPAM: Cannot get cache class %d, type: %d\n",
+                   cache_level, cache_type);
+            spin_unlock(&mpam_list_lock);
+            return;
+        }
+    }
+    spin_unlock(&mpam_list_lock);
+
+    *size = cache->size;
+}
+
 /*
  * Local variables:
  * mode: C
diff --git a/xen/arch/arm/sysctl.c b/xen/arch/arm/sysctl.c
index b441f6a80eb3..55d1ca080974 100644
--- a/xen/arch/arm/sysctl.c
+++ b/xen/arch/arm/sysctl.c
@@ -11,6 +11,7 @@
 #include <xen/lib.h>
 #include <xen/dt-overlay.h>
 #include <xen/errno.h>
+#include <xen/guest_access.h>
 #include <xen/hypercall.h>
 #include <asm/arm64/sve.h>
 #include <public/sysctl.h>
@@ -39,12 +40,63 @@ long arch_do_sysctl(struct xen_sysctl *sysctl,
     case XEN_SYSCTL_psr_cmt_op:
         if ( !cpu_has_mpam )
             return -EOPNOTSUPP;
+
+        if ( sysctl->u.psr_cmt_op.flags != 0 )
+            return -EINVAL;
+
+        switch ( sysctl->u.psr_cmt_op.cmd )
+        {
+        case XEN_SYSCTL_PSR_CMT_get_total_rmid:
+            sysctl->u.psr_cmt_op.u.data = mpam_get_system_max_pmg();
+            break;
+        case XEN_SYSCTL_PSR_CMT_get_l3_cache_size:
+        {
+            unsigned int l3_size = 0;
+
+            mpam_get_cache_size(MPAM_CLASS_CACHE, 3, &l3_size);
+            /* in KB unit */
+            sysctl->u.psr_cmt_op.u.data = (l3_size ? l3_size / 1024 : 0);
+
+            break;
+        }
+        default:
+            sysctl->u.psr_cmt_op.u.data = 0;
+            ret = -EOPNOTSUPP;
+            break;
+        }
+
+        if ( __copy_to_guest(u_sysctl, sysctl, 1) )
+            ret = -EFAULT;
         break;
 
     case XEN_SYSCTL_psr_alloc:
+    {
+        unsigned int hw_val;
+
         if ( !cpu_has_mpam )
             return -EOPNOTSUPP;
+
+        switch ( sysctl->u.psr_alloc.cmd )
+        {
+        case XEN_SYSCTL_PSR_get_l3_info:
+            ret = mpam_get_hwinfo(MPAM_CLASS_CACHE, 3, &hw_val);
+            if( ret )
+                break;
+
+            sysctl->u.psr_alloc.u.cat_info.cos_max =
+                mpam_get_system_max_partid();
+            sysctl->u.psr_alloc.u.cat_info.cbm_len = hw_val;
+
+            if ( __copy_field_to_guest(u_sysctl, sysctl, u.psr_alloc) )
+                ret = -EFAULT;
+            break;
+        default:
+            ret = -EOPNOTSUPP;
+            break;
+        }
+
         break;
+    }
 #endif
 
     default:
-- 
2.34.1

